src/TestExplorer/LSPTestDiscovery.ts

```diff
@@ -29,104 +34,118 @@ import { LanguageClientManager } from "../sourcekit-lsp/LanguageClientManager";
-            const testsInWorkspace = tests.filter(item =>
-                isPathInsidePath(
-                    client.protocol2CodeConverter.asLocation(item.location).uri.fsPath,
-                    workspaceRoot.fsPath
-                )
-            );
-            const classes = testsInWorkspace
-                .filter(item => {
-                    return (
-                        item.kind === langclient.SymbolKind.Class &&
-                        isPathInsidePath(
-                            client.protocol2CodeConverter.asLocation(item.location).uri.fsPath,
-                            workspaceRoot.fsPath
-                        )
-                    );
-                })
-                .map(item => {
-                    const functions = testsInWorkspace
-                        .filter(func => func.containerName === item.name)
-                        .map(func => {
-                            const openBrackets = func.name.indexOf("(");
-                            let funcName = func.name;
-                            if (openBrackets) {
-                                funcName = func.name.slice(0, openBrackets);
-                            }
-                            return {
-                                name: funcName,
-                                location: client.protocol2CodeConverter.asLocation(func.location),
-                            };
-                        });
-                    return {
-                        name: item.name,
-                        location: client.protocol2CodeConverter.asLocation(item.location),
-                        functions: functions,
-                    };
-                });
-            return classes;
+            // Only use the lsp for this request if it supports the
+            // workspace/tests method, and is at least version 2.
+            if (this.checkExperimentalCapability(client, workspaceTestsRequest.method, 2)) {
+                const tests = await client.sendRequest(workspaceTestsRequest, {}, token);
+                const testsInWorkspace = tests.filter(item =>
+                    isPathInsidePath(
+                        client.protocol2CodeConverter.asLocation(item.location).uri.fsPath,
+                        workspaceRoot.fsPath
+                    )
+                );
+
+                return this.transform(client, swiftPackage, testsInWorkspace);
+            } else {
+                throw new Error(`${workspaceTestsRequest.method} requests not supported`);
+            }
+        });
+    }
+
+    /**
+     * Returns `true` if the LSP supports the supplied `method` at or
+     * above the supplied `minVersion`.
+     */
+    private checkExperimentalCapability(
+        client: LanguageClient,
+        method: string,
+        minVersion: number
+    ) {
+        const experimentalCapability = client.initializeResult?.capabilities.experimental;
+        if (!experimentalCapability) {
```

Comment:
I did not know about this. It would be good if Microsoft actually documented the language client APIs. Is the availability of the workspace/tests and textDocument/tests something we can cache?

Reply:
The capabilities are actually cached on the `client` right after the connection is established to the server.

Reply:
I actually meant caching in a boolean, instead of running this function. It's a small amount of code, but it gets run with every call

Reply:
Ah ok, I've cached the result in a Map keyed off the capability name + version.